lib/vrl/compiler/src/expression/block.rs

```diff
@@ -49,21 +49,36 @@ impl Expression for Block {
         last.resolve(ctx)
     }
 
+    /// If an expression has a "never" type, it is considered a "terminating" expression.
+    /// Type information of future expressions in this block should not be considered after
+    /// a terminating expression.
+    ///
+    /// Since type definitions due to assignments are calculated outside of the "type_def" function,
+    /// assignments that can never execute might still have adjusted the type definition.
+    /// Therefore, expressions after a terminating expression must not be included in a block.
+    /// It is considered an internal compiler error if this situation occurs, which is checked here
+    /// and will result in a panic.
+    ///
+    /// VRL is allowed to have expressions after a terminating expression, but the compiler
+    /// MUST not include them in a block expression when compiled.
     fn type_def(&self, (_, external): (&LocalEnv, &ExternalEnv)) -> TypeDef {
-        let mut type_defs = self
-            .inner
-            .iter()
-            .map(|expr| expr.type_def((&self.local_env, external)))
-            .collect::<Vec<_>>();
-
-        // If any of the stored expressions is fallible, the entire block is
-        // fallible.
-        let fallible = type_defs.iter().any(TypeDef::is_fallible);
-
-        // The last expression determines the resulting value of the block.
-        let type_def = type_defs.pop().unwrap_or_else(TypeDef::null);
+        let mut last = TypeDef::null();
+        let mut fallible = false;
+        let mut has_terminated = false;
+        for expr in &self.inner {
+            if has_terminated {
+                panic!("VRL block contains an expression after a terminating expression. This is an internal compiler error. Please submit a bug report.");
```

Comment:
I like this 👍 We should get in the habit of raising these explicit panics, instead of using `unreachable` with a cryptic internal comment.